File: v16.1/BuildEntityRelations.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

class BuildEntityRelations
{
    static void Main(string[] args)
    {
        string dataDir = args.Length > 0 ? args[0] : "data";
        string tablesPath = Path.Combine(dataDir, "tables.json");
        string outputPath = args.Length > 1 ? args[1] : Path.Combine(dataDir, "entity-relations.json");

        Console.Error.WriteLine("Reading {0}...", tablesPath);
        var root = JObject.Parse(File.ReadAllText(tablesPath));
        var tables = (JObject)root["tables"];

        // Build entity list and relationship list
        var entities = new JObject();
        var relationships = new JArray();

        // Track relationship names for uniqueness
        var relSet = new HashSet<string>();

        foreach (var prop in tables.Properties())
        {
            string tableName = prop.Name;
            var tbl = (JObject)prop.Value;

            // Build entity
            var entity = new JObject();

            // Primary key
            entity["primaryKey"] = tbl["primaryKey"];

            // Columns: just name and type
            var cols = new JArray();
            foreach (var c in (JArray)tbl["columns"])
            {
                var col = new JObject();
                col["name"] = c["name"];
                col["type"] = c["type"];
                col["nullable"] = c["nullable"];
                if (c["identity"] != null && (bool)c["identity"])
                    col["identity"] = true;
                cols.Add(col);
            }
            entity["columns"] = cols;

            entities[tableName] = entity;

            // Build relationships from foreign keys
            var fks = (JArray)tbl["foreignKeys"];
            if (fks != null)
            {
                foreach (var fk in fks)
                {
                    string fromCol = (string)fk["column"];
                    string refStr = (string)fk["references"];
                    string[] parts = refStr.Split('.');
                    string toTable = parts[0];
                    string toCol = parts[1];

                    // Deduplicate
                    string key = tableName + "." + fromCol + "->" + toTable + "." + toCol;
                    if (relSet.Contains(key))
                        continue;
                    relSet.Add(key);

                    var rel = new JObject();
                    rel["fromTable"] = tableName;
                    rel["fromColumn"] = fromCol;
                    rel["toTable"] = toTable;
                    rel["toColumn"] = toCol;

                    // Determine cardinality:
                    // FK column in the source table pointing to PK in target = many-to-one
                    rel["type"] = "many-to-one";

                    relationships.Add(rel);
                }
            }
        }

        Console.Error.WriteLine("Entities: {0}, Relationships: {1}", entities.Count, relationships.Count);

        // Build output
        var output = new JObject();
        var metadata = new JObject();
        metadata["database"] = (string)root["database"];
        metadata["server"] = (string)root["server"];
        metadata["version"] = "16.1";
        metadata["generated"] = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
        metadata["entityCount"] = entities.Count;
        metadata["relationshipCount"] = relationships.Count;
        output["metadata"] = metadata;
        output["entities"] = entities;
        output["relationships"] = relationships;

        File.WriteAllText(outputPath, output.ToString(Formatting.Indented));
        Console.Error.WriteLine("Written to {0}", outputPath);
    }
}

File: v16.1/build_entity_relations.py
"""Build entity-relations.json from tables.json for the ER Explorer web app."""

import json
import sys
from datetime import datetime
from pathlib import Path


def build_entity_relations(tables_data):
    """Transform tables.json data into entity-relations.json format."""
    tables = tables_data["tables"]
    entities = {}
    relationships = []
    rel_set = set()

    for table_name, tbl in tables.items():
        # Build entity
        entity = {
            "primaryKey": tbl.get("primaryKey", []),
            "columns": [
                {
                    "name": c["name"],
                    "type": c["type"],
                    "nullable": c["nullable"],
                    **({"identity": True} if c.get("identity") else {}),
                }
                for c in tbl.get("columns", [])
            ],
        }
        entities[table_name] = entity

        # Build relationships from foreign keys
        for fk in tbl.get("foreignKeys", []):
            from_col = fk["column"]
            ref_parts = fk["references"].split(".")
            to_table = ref_parts[0]
            to_col = ref_parts[1]

            key = f"{table_name}.{from_col}->{to_table}.{to_col}"
            if key in rel_set:
                continue
            rel_set.add(key)

            relationships.append({
                "fromTable": table_name,
                "fromColumn": from_col,
                "toTable": to_table,
                "toColumn": to_col,
                "type": "many-to-one",
            })

    print(f"Entities: {len(entities)}, Relationships: {len(relationships)}", file=sys.stderr)

    return {
        "metadata": {
            "database": tables_data.get("database", ""),
            "server": tables_data.get("server", ""),
            "version": "16.1",
            "generated": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            "entityCount": len(entities),
            "relationshipCount": len(relationships),
        },
        "entities": entities,
        "relationships": relationships,
    }


def main(data_dir=None, output_path=None):
    if data_dir is None:
        data_dir = Path(__file__).resolve().parent / "data"
    data_dir = Path(data_dir)

    tables_path = data_dir / "tables.json"
    print(f"Reading {tables_path}...", file=sys.stderr)
    tables_data = json.loads(tables_path.read_text(encoding="utf-8-sig"))

    result = build_entity_relations(tables_data)

    if output_path is None:
        output_path = data_dir / "entity-relations.json"
    output_path = Path(output_path)
    output_path.write_text(json.dumps(result, indent=2, ensure_ascii=False), encoding="utf-8")
    print(f"Written to {output_path}", file=sys.stderr)
    return result


if __name__ == "__main__":
    d = sys.argv[1] if len(sys.argv) > 1 else None
    o = sys.argv[2] if len(sys.argv) > 2 else None
    main(d, o)
